### PR TITLE
Convert the handwriting recognition example to Keras 3.

### PR DESCRIPTION
**Pull request by fkouteib**

### CODE REVIEW

File: examples/vision/handwriting_recognition.py [hunk 1]
```diff
@@ -338,91 +349,95 @@ def prepare_dataset(image_paths, labels):
...
-
-        input_length = input_length * tf.ones(shape=(batch_len, 1), dtype="int64")
-        label_length = label_length * tf.ones(shape=(batch_len, 1), dtype="int64")
-        loss = self.loss_fn(y_true, y_pred, input_length, label_length)
-        self.add_loss(loss)
-
-        # At test time, just return the computed predictions.
-        return y_pred
+# Ported from keras 2 backend (with keras 3 ops updates).
+def ctc_decode(y_pred, input_length, greedy=True, beam_width=100, top_paths=1):
```
Comment (fchollet) [inline examples/vision/handwriting_recognition.py:358]: In the future you should be able to use ops from https://github.com/keras-team/keras/pull/19366

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. The code looks good so far -- do you need any help?

**Comment by fkouteib** [comment]
Hi @fchollet Thanks for the review. I am currently having issues with loading the dataset followed by an early crash for lack of input. 

> Total training samples: 86810
Total validation samples: 4823
Total test samples: 4823
Maximum length:  21
Vocab size:  78
2024-04-15 01:12:03.619746: W tensorflow/core/kernels/data/cache_dataset_ops.cc:858] **The calling iterator did not fully read the dataset being cached.** In order to avoid unexpected truncation of the dataset, the partially cached contents of the dataset  will be discarded. This can happen if you have an input pipeline similar to `dataset.cache().take(k).repeat()`. You should use `dataset.take(k).cache().repeat()` instead.
2024-04-15 01:12:03.755489: W tensorflow/core/framework/local_rendezvous.cc:404] **Local rendezvous is aborting with status: OUT_OF_RANGE: End of sequence**
Model: "handwriting_recognizer"
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                  ┃ Output Shape              ┃         Param # ┃ Connected to               ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ image (InputLayer)            │ (None, 128, 32, 1)        │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ Conv1 (Conv2D)                │ (None, 128, 32, 32)       │             320 │ image[0][0]                │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ pool1 (MaxPooling2D)          │ (None, 64, 16, 32)        │               0 │ Conv1[0][0]                │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ Conv2 (Conv2D)                │ (None, 64, 16, 64)        │          18,496 │ pool1[0][0]                │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ pool2 (MaxPooling2D)          │ (None, 32, 8, 64)         │               0 │ Conv2[0][0]                │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ reshape (Reshape)             │ (None, 32, 512)           │               0 │ pool2[0][0]                │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ dense1 (Dense)                │ (None, 32, 64)            │          32,832 │ reshape[0][0]              │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ dropout (Dropout)             │ (None, 32, 64)            │               0 │ dense1[0][0]               │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ bidirectional (Bidirectional) │ (None, 32, 256)           │         197,632 │ dropout[0][0]              │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ bidirectional_1               │ (None, 32, 128)           │         164,352 │ bidirectional[0][0]        │
│ (Bidirectional)               │                           │                 │                            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ label (InputLayer)            │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ softmax (Dense)               │ (None, 32, 81)            │          10,449 │ bidirectional_1[0][0]      │
└───────────────────────────────┴───────────────────────────┴─────────────────┴────────────────────────────┘
 Total params: 424,081 (1.62 MB)
 Trainable params: 424,081 (1.62 MB)
 Non-trainable params: 0 (0.00 B)
2024-04-15 01:12:04.273326: W tensorflow/core/framework/local_rendezvous.cc:404] Local rendezvous is aborting with status: OUT_OF_RANGE: End of sequence
Traceback (most recent call last):
  File "/home/fkouteib/devspace/keras-io/examples/vision/handwriting_recognition.py", line 519, in <module>
    prediction_model = keras.models.Model(
  File "/home/fkouteib/.keras_venv/lib/python3.10/site-packages/keras/src/models/model.py", line 141, in __new__
    return functional.Functional(*args, **kwargs)
  File "/home/fkouteib/.keras_venv/lib/python3.10/site-packages/keras/src/utils/tracking.py", line 28, in wrapper
    return fn(*args, **kwargs)
  File "/home/fkouteib/.keras_venv/lib/python3.10/site-packages/keras/src/models/functional.py", line 162, in __init__
    Function.__init__(self, inputs, outputs, name=name, **kwargs)
  File "/home/fkouteib/.keras_venv/lib/python3.10/site-packages/keras/src/ops/function.py", line 62, in __init__
    raise ValueError(
**ValueError: `inputs` argument cannot be empty. Received:
inputs=[]**
outputs=<KerasTensor shape=(None, 32, 81), dtype=float32, sparse=False, name=keras_tensor_19>

When I change 
`return dataset.batch(batch_size).cache.prefetch(AUTOTUNE)`
to
`return dataset.batch(batch_size).prefetch(AUTOTUNE)`

The local rendezvous abort error and subsequent early exit persist. Have you encountered either of these issues before? any debug pointers are appreciated. Thanks.

**Comment by fchollet** [comment]
Did you take a look at the data on local disk and checked everything looked good?

**Review by fchollet — COMMENTED** [review]
Looking good! Does it work with all backends or does it still require TF?

**Comment by fkouteib** [comment]
It's still TF only.  I still have the following TF dependencies unless agnostic Keras equivalents exist that I am not aware of.

* tf.data.Dataset() // I think works across all Keras backends.
* tf.io.read_file()
* tf.image.decode_png()
* tf.strings..unicode_split()
* tf.strings.reduce_join()
* tf.sparse.from_dense()

Of these, the biggest blocker seems like support for sparse tensors in Keras with non-TF backend. I found https://github.com/keras-team/keras/issues/18420 but don't know the current status.

**Pull request closed without merging**